### PR TITLE
:octocat: Porting CNN, TF ➡️  PyTorch 

### PR DESCRIPTION
**Pull request by shivance**
This PR adds `CNN` class and adds suitable unit tests.

# Description
## What are we doing here?
============================
`CNN` class is PyTorch translation of deepchem's Keras CNN implementation. 
Removing the old Tensorflow CNN implementation from deepchem.

## Why are we doing this ?
Convolutional Neural ODE is the proposed project for deepchem under `Deepchem Open Source Fellowship` .  It requires a convolutional encoder. It will be a good practice to implement Torch Version of CNN which further can be reused for other purposes (given that Deepchem is already moving towards PyTorch ecosystem)

### CODE REVIEW

File: deepchem/models/torch_models/cnn.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+from typing import List, Union, Any, Type
+
+try:
+  from collections.abc import Sequence as SequenceCollection
+except:
+  from collections import Sequence as SequenceCollection
+
+
+class TorchCNN(nn.Module):
+  """A 1, 2, or 3 dimensional convolutional network for either regression or classification.
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:15]: This should numpydoc style

https://numpydoc.readthedocs.io/en/latest/format.html

File: deepchem/models/torch_models/cnn.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+               dropouts: Union[float, List[float]] = 0.5,
+               activation_fns=nn.ReLU,
+               pool_type: str = 'max',
+               mode: str = 'classification',
+               n_classes: int = 2,
+               uncertainty: bool = False,
+               residual: bool = False,
+               padding: Union[int, str] = 'valid') -> None:
+    """Create a CNN.
+        Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:57]: I think you're mixing tabs and spaces here

File: deepchem/models/torch_models/cnn.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+      if activation_fn is not None:
+        self.layers.append(activation_fn())
+
+      self.layers.append(PoolLayer(size))
+      in_shape = out_shape
+
+  def forward(self, x: torch.Tensor) -> List[Any]:
+    """
+        Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:185]: Same formatting issue

File: deepchem/models/torch_models/cnn.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    for layer in self.layers:
+      x = layer(x)
+      # residual blocks can only be used when successive layers have the same output shape
+      if self.residual and layer.in_channels == layer.out_channels:
+        x = x + prev_layer
+      prev_layer = x
+
+    outputs = []
+    batch_size = x.shape[0]
+
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:206]: Can you add a comment explaining the use of einops and what the rearrangement here does?

File: deepchem/models/tests/test_torch_cnn.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+try:
+  import torch
+  from deepchem.models import TorchCNN
+  has_pytorch = True
+except:
+  has_pytorch = False
+
+
+@pytest.mark.torch
+def test_cnn_torch():
```
Comment (rbharath) [inline deepchem/models/tests/test_torch_cnn.py:11]: Have tests for 1D, 2D, 3D

File: setup.py [hunk 6]
```diff
@@ -37,45 +37,40 @@ def _get_version():
...
-        'Intended Audience :: Information Technology',
-        'License :: OSI Approved :: MIT License',
-        'Operating System :: OS Independent',
-        'Programming Language :: Python :: 3.7',
-        'Programming Language :: Python :: 3.8',
-        'Programming Language :: Python :: 3.9',
-    ],
-    license='MIT',
-    description='Deep learning models for drug discovery, \
+setup(name='deepchem',
```
Comment (rbharath) [inline setup.py:40]: This file should not be modified. We should add this as a soft dependency and not a hard dependency.

Ideally, can we avoid use of einops? Adding an extra dependency to deepchem is always extra maintenance burden. einops is fairly popular so it may be OK, but good to see if we can avoid it

Comment (rbharath) [reply setup.py:40]: If a new soft dependency is added, it will need to be added to the docs/ folder as well

Comment (peastman) [reply setup.py]: [`rearrange()`](https://einops.rocks/api/rearrange/) seems to be nothing more than a different API for basic numpy operations like transpose and reshape.  We should be able to easily do the same thing with numpy calls.  It's not worth adding a new dependency for that.

**Comment by peastman** [comment]
None of the tests actually verifies that the model is working.  It just invokes an untrained model and checks whether the output has the right shape, which doesn't tell us whether it's correct.  Compare to [test_cnn.py](https://github.com/deepchem/deepchem/blob/master/deepchem/models/tests/test_cnn.py), which fits some regression and classification models and verifies they can reproduce the training data.

**Comment by peastman** [comment]
Rather than keeping around the TensorFlow based CNN class and adding a new TorchCNN class, I think it would be better to convert the existing CNN class to PyTorch, keeping the whole API as close to unchanged as possible.  Hardly anything in the API is specific to TensorFlow, so users have little reason to care which library the model is implemented with.  We're converting it for our own benefit, because we think PyTorch will be a better foundation to build on going forward.

**Comment by shivance** [comment]
> 

Currently working on TorchModel wrapper until then @rbharath asked to get the basic model merged first. The wrapper will enable the fit method to be called on the model

**Comment by shivance** [comment]
@peastman That's fascinating ! I didn't try reshaping torch tensor using numpy. It works . But here is something : 

```
pattern = ("b c h -> b (c h)", "b c h w -> b (c h w)", "b c h w a -> b (c h w a)")[self.dims - 1]
x = rearrange(x, pattern)
```

We'll be using :
`x = np.reshape(x,(batch_size, -1))`

but it yields `Error : Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead`

using,
`x = np.reshape(x.detach(), (batch_size, -1)` 
reshapes it to a `torch.Tensor without gradient but required shape` without Error , but it's a Torch Tensor which requires gradient !

So I think following should do the job  
```
 x = np.reshape(x.detach(),(batch_size, -1))
 x.requires_grad = True
```

Shouldn't it ?

**Comment by peastman** [comment]
Since `x` is a tensor, use `torch.reshape()` instead of `np.reshape()`.

**Review by peastman — COMMENTED** [review]
I still think it would be best to remove the old TensorFlow CNN model.  At the moment there are two different CNN models, each with a different test suite.  There's no need for that.  And part of the proof of the new model is that the old test suite continues to pass with the new implementation.

**Comment by rbharath** [comment]
+1 for @peastman's comment.

@shivance Can you go ahead and remove the old model but verify that old tests continue to pass with the new implementation?

**Comment by shivance** [comment]
@peastman @rbharath 
[1] The tests failing. Says Module not found error.. even though I have fixed imports
[2] For my implementation : 
          For classification, `roc_auc_score` is always < 0.9 while for regression `mean_squared_error` is always > 0.1 
      How can we improve the scores and match to the standard of existing model ?

**Comment by peastman** [comment]
>  For classification, roc_auc_score is always < 0.9 while for regression mean_squared_error is always > 0.1

If the new model can't fit a small dataset, that means you have a bug somewhere!

**Comment by peastman** [comment]
Why are you creating a new layer inside `forward()`?  That's certainly not the correct thing to do.

Look at MultitaskClassifier for an example of how another classification model handles it.

File: deepchem/models/torch_models/cnn.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    x = torch.reshape(x, (batch_size, -1))
+
+    if self.mode == "classification":
+
+      logits = nn.Linear(x.shape[1], self.n_tasks * self.n_classes)(x)
```
Comment (shivance) [inline deepchem/models/torch_models/cnn.py:219]: @peastman, Let's discuss about the solution to root issue, If we are to define this nn.Linear in constructor, how can we get the input shape which is x.shape[1] ?
pytorch defines graphs dynamically , which means we can get it only at runtime

I'm trying to replicate your [TF Implementation in torch](https://github.com/deepchem/deepchem/blob/66116f1b8edd8130eff0a435c8cb8d3deefaab75/deepchem/models/cnn.py#L205)

**Comment by shivance** [comment]
I figured out the solution [here](https://discuss.pytorch.org/t/how-to-initialize-feed-forward-network-in-multitask-cnn/156469). Used this in code, and the scores now are really good ! @peastman @rbharath

File: deepchem/models/torch_models/cnn.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+                        stride=stride,
+                        padding=padding,
+                        dilation=1,
+                        groups=1,
+                        bias=True)
+
+      nn.init.normal_(layer.weight, 0, weight_stddev)
+
+      layer_bias: torch.Tensor = layer.bias
+      nn.init.constant_(layer_bias, bias_const)
```
Comment (tonydavis629) [inline deepchem/models/torch_models/cnn.py:208]: Why not use layer.bias in nn.init.constant_ rather than define a new variable layer_bias?

Comment (shivance) [reply deepchem/models/torch_models/cnn.py:208]: I did, it was giving the same CI failure, (mypy check running fine on local though), 
After which i tried typecasting, which in turn gave the same failure.
Didn't change it afterwards

Comment (tonydavis629) [reply deepchem/models/torch_models/cnn.py:208]: What version of mypy are you using locally?

Comment (shivance) [reply deepchem/models/torch_models/cnn.py:208]: The latest available through pip

File: deepchem/models/tests/test_reload.py [hunk 9]
```diff
@@ -924,96 +905,39 @@ def test_textCNN_classification_reload():
   assert len(model.model.layers) == len(reloaded_model.model.layers)
 
 
-@pytest.mark.tensorflow
```
Comment (rbharath) [inline deepchem/models/tests/test_reload.py:927]: This should remain unchanged

**Comment by shivance** [comment]
@peastman I think the PR is gtg as of now. Can you review it once more?

File: deepchem/models/cnn.py [hunk 10]
```diff
@@ -1,259 +0,0 @@
...
-from tensorflow.keras.layers import Input, Dense, Reshape, Softmax, Dropout, Activation, Lambda
-import tensorflow.keras.layers as layers
-try:
-  from collections.abc import Sequence as SequenceCollection
-except:
-  from collections import Sequence as SequenceCollection
-
-
-class CNN(KerasModel):
-  """A 1, 2, or 3 dimensional convolutional network for either regression or classification.
```
Comment (tonydavis629) [inline deepchem/models/cnn.py:16]: So are we removing TF CNN capability? Should we maybe create a keras_models module for the TF CNN?

Comment (shivance) [reply deepchem/models/cnn.py:16]: yes, as suggested by @peastman 

> I still think it would be best to remove the old TensorFlow CNN model. At the moment there are two different CNN models, each with a different test suite. There's no need for that. And part of the proof of the new model is that the old test suite continues to pass with the new implementation.

File: deepchem/models/__init__.py [hunk 11]
```diff
@@ -48,6 +47,7 @@
   from deepchem.models.torch_models import Pagtn, PagtnModel
   from deepchem.models.fcnet import MultitaskRegressor, MultitaskClassifier, MultitaskFitTransformRegressor
   from deepchem.models.torch_models import MEGNetModel
+  from deepchem.models.torch_models import CNN, CNNModule
```
Comment (rbharath) [inline deepchem/models/__init__.py:50]: Don't export `CNNModule` here. It's a layer and we can export it elsewhere

File: docs/source/api_reference/models.rst [hunk 12]
```diff
@@ -266,6 +266,12 @@ WGAN
 .. autoclass:: deepchem.models.WGAN
   :members:
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:268]: `CNN` should be moved to the torch model section of the doc (https://github.com/deepchem/deepchem/blob/master/docs/source/api_reference/models.rst#pytorch-models) from the existing keras model

File: docs/source/api_reference/models.rst [hunk 13]
```diff
@@ -266,6 +266,12 @@ WGAN
...
 
+CNNModule
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:269]: CNNModule should be added to https://github.com/deepchem/deepchem/blob/master/docs/source/api_reference/layers.rst

Comment (rbharath) [reply docs/source/api_reference/models.rst:269]: https://github.com/deepchem/deepchem/blob/master/docs/source/api_reference/layers.rst#torch-layers

File: deepchem/models/__init__.py (hunk 11 above)
Comment (rbharath) [reply deepchem/models/__init__.py:50]: Export `CNNModule` from `deepchem.models.torch_models.layers`

File: deepchem/models/torch_models/cnn.py [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+        var = torch.exp(log_var)
+        outputs = [output, var, output, log_var]
+
+      else:
+        outputs = [output]
+
+    return outputs
+
+
+class CNN(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:278]: Copy over the old docstring from `CNN` here and make sure to add a usage example

File: deepchem/models/torch_models/cnn.py [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+
+    loss: Union[L2Loss, Callable[[Any, Any, Any], Any]]
+
+    if uncertainty:
+
+      def loss(outputs, labels, weights):
+
+        diff = labels[0] - outputs[0]
+
+        return torch.mean(diff**2 / torch.exp(outputs[1]) + outputs[1])
```
Comment (tonydavis629) [inline deepchem/models/torch_models/cnn.py:404]: I don't think this should be double spaced

Comment (shivance) [reply deepchem/models/torch_models/cnn.py]: Sure. I've removed line spacing within the function

File: deepchem/models/torch_models/cnn.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+  >>> n_samples = 10
+  >>> n_features = 3
+  >>> n_tasks = 1
+  >>> np.random.seed(123)
+  >>> X = np.random.rand(n_samples, 10, n_features)
+  >>> y = np.random.randint(2, size=(n_samples, n_tasks)).astype(np.float32)
+  >>> dataset: dc.data.Dataset = dc.data.NumpyDataset(X, y)
+  >>> regression_metric = dc.metrics.Metric(dc.metrics.mean_squared_error)
+  >>> model = CNN(n_tasks, n_features, dims=1, kernel_size=3, mode='regression')
+  >>> avg_loss = model.fit(dataset, nb_epoch=200)
```
Comment (rbharath) [inline deepchem/models/torch_models/cnn.py:42]: Can you make this 10 epochs instead

**Review by rbharath — COMMENTED** [review]
I think this is ready to merge in from my end.

@peastman @tonydavis629 Do either of you have additional comments on this PR that should be addressed before merge?

**Comment by peastman** [comment]
Looks good to me.

**Review by rbharath — APPROVED** [review]
Ok going to go ahead and approve since @peastman has had a chance to take a look.

Nice work @shivance!
